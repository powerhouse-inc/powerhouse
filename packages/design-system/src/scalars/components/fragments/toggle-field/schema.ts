import { z } from "zod";

export const formSchema = z
  .object({
    checked: z.boolean().optional(),
    required: z.boolean().optional(),
  })
  .refine(
    (data) => {
      if (data.checked && data.required !== true) {
        return false;
      }
      return true;
    },
    {
      message: "This field its requiere",
      path: ["checked"],
    },
  );

export const validateRequiredField = (required: boolean, checked: boolean) => {
  const validations = formSchema.safeParse({ required, checked });
  return validations.success
    ? []
    : validations.error.errors.map((e) => e.message);
};
