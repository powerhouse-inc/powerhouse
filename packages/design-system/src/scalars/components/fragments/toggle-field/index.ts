export { ToggleField } from "./toggle-field.js";
