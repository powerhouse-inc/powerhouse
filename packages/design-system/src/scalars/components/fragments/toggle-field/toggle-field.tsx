import {
  Toggle,
  type ToggleProps,
} from "../../../../ui/components/data-entry/toggle/toggle.js";
import { withFieldValidation } from "../with-field-validation/with-field-validation.js";

const ToggleField = withFieldValidation<ToggleProps>(Toggle);

ToggleField.displayName = "ToggleField";

export { ToggleField };
