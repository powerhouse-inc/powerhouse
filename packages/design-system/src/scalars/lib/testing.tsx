import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
