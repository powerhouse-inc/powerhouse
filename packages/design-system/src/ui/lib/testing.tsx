import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
