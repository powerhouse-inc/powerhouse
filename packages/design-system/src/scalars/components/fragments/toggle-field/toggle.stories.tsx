import { getDefaultArgTypes, getValidationArgTypes, withForm } from "#scalars";
import type { Meta, StoryObj } from "@storybook/react";
import { ToggleField } from "./toggle-field";

const meta: Meta<typeof ToggleField> = {
  title: "Document Engineering/Fragments/Toggle Field",
  tags: ["autodocs"],
  component: ToggleField,
  decorators: [withForm],
  parameters: {
    chromatic: {
      disableSnapshot: true,
    },
  },
  argTypes: {
    ...getDefaultArgTypes({
      valueControlType: "boolean",
      valueType: "boolean",
    }),
    ...getValidationArgTypes({
      enabledArgTypes: {
        validators: false,
        showErrorOnBlur: false,
        showErrorOnChange: false,
      },
    }),
  },
  args: {
    name: "toggle",
    errors: [],
    warnings: [],
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const DefaultChecked: Story = {
  args: {
    value: true,
  },
};

export const Unchecked: Story = {
  name: "Unchecked",
  args: {
    value: false,
  },
};

export const CheckedWithLabel: Story = {
  name: "Checked with label",
  args: {
    value: true,
    label: "Active",
  },
};

export const UncheckedWithLabel: Story = {
  name: "Unchecked with label",
  args: {
    value: false,
    label: "Active",
  },
};

export const DisabledChecked: Story = {
  name: "Disabled checked",
  args: {
    value: true,
    disabled: true,
  },
};

export const DisabledUncheckedWithoutLabel: Story = {
  name: "Disabled unchecked",
  args: {
    value: false,
    disabled: true,
  },
};
