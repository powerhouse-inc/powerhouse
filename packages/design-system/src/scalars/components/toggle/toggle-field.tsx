import React, { useId } from "react";
import { Toggle } from "./toggle";
import { cn } from "@/scalars/lib/utils";
import { FormLabel } from "../form-label";
import { FormMessageList } from "../form-message";

interface ToggleFieldProps {
  label?: string;
  disabled?: boolean;
  checked?: boolean;
  onCheckedChange?: (checked: boolean) => void;
  errors?: string[];
  className?: string;
  required?: boolean;
  name?: string;
}

const ToggleField: React.FC<ToggleFieldProps> = ({
  label,
  disabled = false,
  checked = true,
  onCheckedChange,
  errors = [],
  required = false,
  name,
  className,
}) => {
  const id = useId();

  return (
    <div
      className={cn("flex flex-col gap-1", className)}
      data-testid="custom-class"
    >
      <div className="flex items-center">
        <Toggle
          aria-labelledby={`${id}-label`}
          required={required}
          disabled={disabled}
          name={name}
          id={id}
          checked={checked}
          onCheckedChange={onCheckedChange}
        />
        {label && (
          <FormLabel
            htmlFor={id}
            className="ml-2"
            disabled={disabled}
            id={`${id}-label`}
          >
            {label}
          </FormLabel>
        )}
      </div>
      {errors.length !== 0 && (
        <FormMessageList messages={errors} type="error" />
      )}
    </div>
  );
};

export { ToggleField };
