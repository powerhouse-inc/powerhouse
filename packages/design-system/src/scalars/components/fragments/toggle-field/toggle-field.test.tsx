import { fireEvent, screen, waitFor } from "@testing-library/react";
import { describe, it, vi } from "vitest";
import { renderWithForm } from "../../../lib/testing.js";
import { ToggleField } from "./toggle-field.js";

describe("ToggleField Component", () => {
  const mockOnChange = vi.fn();

  it("should match snapshot", () => {
    const { container } = renderWithForm(<ToggleField name="test" />);
    expect(container).toMatchSnapshot();
  });

  it("should render default status without a label on the left", () => {
    renderWithForm(<ToggleField name="test" />);
    expect(screen.queryByText("Test Label")).not.toBeInTheDocument();
  });

  it("should render with a label when label prop is provided", () => {
    renderWithForm(<ToggleField name="test" label="Test Label" />);
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should render checked status without label", () => {
    renderWithForm(<ToggleField name="test" value={true} />);
    expect(screen.queryByText("Test Label")).not.toBeInTheDocument();
  });

  it("should render checked status with a label on the left", () => {
    renderWithForm(<ToggleField name="test" label="Test Label" value={true} />);
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should not render the label when not provided", () => {
    renderWithForm(<ToggleField name="test" />);
    expect(screen.queryByText("Test Label")).not.toBeInTheDocument();
  });

  it("should display an error message when hasMessage is true", async () => {
    renderWithForm(
      <ToggleField name="test" label="Test Label" errors={["Error message"]} />,
    );
    await waitFor(() =>
      expect(screen.getByText("Error message")).toBeInTheDocument(),
    );
  });

  it("should call onChange when clicked", () => {
    renderWithForm(
      <ToggleField name="test" label="Test Label" onChange={mockOnChange} />,
    );
    const toggleInput = screen.getByRole("switch");

    fireEvent.click(toggleInput);
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(toggleInput).toBeInTheDocument();
  });

  it("should disable the toggle when disabled prop is true", () => {
    renderWithForm(<ToggleField name="test" disabled />);
    const toggle = screen.getByRole("switch");
    expect(toggle).toBeDisabled();
  });

  it("should render with custom className", () => {
    // this is a custom class name for testing purposes
    renderWithForm(<ToggleField name="test" className="custom-class" />);
    const toggle = screen.getByTestId("custom-class");
    expect(toggle).toHaveClass("custom-class");
  });
});
