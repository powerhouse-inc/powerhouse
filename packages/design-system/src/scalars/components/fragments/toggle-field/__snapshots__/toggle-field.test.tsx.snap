// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleField Component > should match snapshot 1`] = `
<div>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-1"
      data-testid="custom-class"
    >
      <div
        class="flex items-center"
      >
        <button
          aria-checked="true"
          aria-labelledby=":r1:-label"
          aria-required="false"
          class="focus-visible:ring-ring focus-visible:ring-offset-background peer inline-flex h-4 w-8 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 data-[state=checked]:bg-[#343839] data-[state=unchecked]:bg-[#D7D8D9] dark:data-[state=checked]:bg-[#3BBE5F] dark:data-[state=unchecked]:bg-[#404446]"
          data-state="checked"
          id=":r1:"
          role="switch"
          type="button"
          value="on"
        >
          <span
            class="peer pointer-events-none block size-3.5 rounded-full bg-[#FCFCFC] shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-3.5 data-[state=unchecked]:translate-x-0"
            data-state="checked"
          />
        </button>
        <input
          aria-hidden="true"
          checked=""
          name="test"
          style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
          tabindex="-1"
          type="checkbox"
          value="on"
        />
      </div>
    </div>
  </form>
</div>
`;
